crates/pipeline-types/src/program_schema.rs

```diff
@@ -0,0 +1,96 @@
+use serde::{Deserialize, Serialize};
+use utoipa::ToSchema;
+
+#[cfg(feature = "testing")]
+use proptest::{collection::vec, prelude::any};
+
+/// Returns canonical form of a SQL identifier:
+/// - If id is _not_ quoted, then it is interpreted as a case-insensitive
+///   identifier and is converted to the lowercase representation
+/// - If id _is_ quoted, then it is a case-sensitive identifier and is returned
+///   as is, without quotes.
+pub fn canonical_identifier(id: &str) -> String {
+    if id.starts_with('"') && id.ends_with('"') && id.len() >= 2 {
+        id[1..id.len() - 1].to_string()
+    } else {
+        id.to_lowercase()
+    }
+}
+
+/// A struct containing the tables (inputs) and views for a program.
+///
+/// Parse from the JSON data-type of the DDL generated by the SQL compiler.
+#[derive(Serialize, Deserialize, ToSchema, Debug, Eq, PartialEq, Clone)]
+#[cfg_attr(feature = "testing", derive(proptest_derive::Arbitrary))]
+pub struct ProgramSchema {
+    #[cfg_attr(
+        feature = "testing",
+        proptest(strategy = "vec(any::<Relation>(), 0..2)")
+    )]
+    pub inputs: Vec<Relation>,
+    #[cfg_attr(
+        feature = "testing",
+        proptest(strategy = "vec(any::<Relation>(), 0..2)")
+    )]
+    pub outputs: Vec<Relation>,
+}
+
+/// A SQL table or view. It has a name and a list of fields.
+///
+/// Matches the Calcite JSON format.
+#[derive(Serialize, Deserialize, ToSchema, Debug, Eq, PartialEq, Clone)]
+#[cfg_attr(feature = "testing", derive(proptest_derive::Arbitrary))]
+pub struct Relation {
+    #[cfg_attr(feature = "testing", proptest(regex = "relation1|relation2|relation3"))]
+    pub name: String,
+    #[serde(default)]
+    pub case_sensitive: bool,
+    #[cfg_attr(feature = "testing", proptest(value = "Vec::new()"))]
+    pub fields: Vec<Field>,
+}
+/// A SQL field.
+///
+/// Matches the Calcite JSON format.
+#[derive(Serialize, Deserialize, ToSchema, Debug, Eq, PartialEq, Clone)]
+#[cfg_attr(feature = "testing", derive(proptest_derive::Arbitrary))]
+pub struct Field {
+    pub name: String,
+    #[serde(default)]
+    pub case_sensitive: bool,
+    pub columntype: ColumnType,
+}
+
+/// A SQL column type description.
+///
+/// Matches the Calcite JSON format.
+#[derive(Serialize, Deserialize, ToSchema, Debug, Eq, PartialEq, Clone)]
+#[cfg_attr(feature = "testing", derive(proptest_derive::Arbitrary))]
+pub struct ColumnType {
+    #[serde(rename = "type")]
+    /// Identifier for the type (e.g., `VARCHAR`, `BIGINT`, `ARRAY` etc.)
+    pub typ: String,
+    /// Does the type accept NULL values?
+    pub nullable: bool,
+    /// Precision of the type.
+    ///
+    /// # Examples
+    /// - `VARCHAR` sets precision to `-1`.
+    /// - `VARCHAR(255)` sets precision to `255`.
+    /// - `BIGINT`, `DATE`, `FLOAT`, `DOUBLE`, `GEOMETRY`, etc. sets precision
+    ///   to None
+    /// - `TIME`, `TIMESTAMP` set precision to `0`.
```

Comment:
This looks wrong, I think that Calcite only works with precision "3" for TIMESTAMP.
Maybe this is a bug we should look into.

Reply:
I just copied this code into a different location, haven't touched it

Reply:
at the time, I verified this by manually creating these types and see what I get back from the compiler